File: src/main/java/net/rainmore/platform/core/models/auth/Role.java
package net.rainmore.platform.core.models.auth;

import net.rainmore.platform.core.models.GenericModel;
import net.rainmore.platform.core.models.Model;

import javax.persistence.*;
import javax.validation.constraints.NotNull;
import java.util.Set;

@Entity
@Table(name = "roles")
public class Role extends GenericModel {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    protected Long id;

    @Column(nullable = false)
    @NotNull
    private String role;

    @ManyToOne(cascade = CascadeType.ALL)
    @JoinTable(name = "users_roles",
            joinColumns = {@JoinColumn(name = "role_id", referencedColumnName = "id")},
            inverseJoinColumns = {@JoinColumn(name = "user_id", referencedColumnName = "id")}
    )
    private Set<User> users;

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getRole() {
        return role;
    }

    public void setRole(String role) {
        this.role = role;
    }

    public Set<User> getUsers() {
        return users;
    }

    public void setUsers(Set<User> users) {
        this.users = users;
    }
}

File: src/main/kotlin/au/com/wytn/centus/domains/core/users/Role.kt
package au.com.wytn.centus.domains.core.users

import au.com.wytn.centus.domains.Model
import javax.persistence.*

@Entity
@Table(name = "userRoles")
class Role : Model {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    var id: Int? = null

    @Column(nullable = false)
    var name: String? = null

    @ManyToOne
    @JoinColumn(name = "parentId")
    var parent: Role? = null

    @Column(nullable = false)
    var isAlmighty: Boolean? = null

    @Column(nullable = false)
    var isArchived: Boolean? = null

    constructor() {}

    constructor(name: String) {
        this.name = name
    }
}